New version adapted to nullable columns and null handling in multi-stage engine (#304)
pinot-contrib/pinot-docs: developers/advanced/null-value-support.md

## Default behavior

Pinot always stores column values in a [forward index](../../basics/indexing/forward-index.md).
Forward index never stores null values but have to store a value for each row.
Therefore independent of the null handling configuration, Pinot always stores a default value for nulls rows in the forward index.
The default value used in a column can be specified in the [schema](../../configuration-reference/schema.md) 

> configuration by setting the `defaultNullValue` field spec.

Reviewer: ```suggestion
The default value used in a column can be specified in the [schema](../../configuration-reference/schema.md) 
configuration by setting the `defaultNullValue` field spec. The `defaultNullValue` depends on the type of data.
```

{% hint style="info" %}
Remember that in the JSON used as table configuration, `defaultNullValue` must always be a String.
If the column type is not String, Pinot will convert that value to the column type automatically.
{% endhint %}

## Store nulls at ingestion time

In order to support any kind of null handling, the first thing that must be done is to store null values in the segments.
As said above, independently on whether null storing is enabled or not, the forward index will always store the default

> value for null rows.

Reviewer: ```suggestion
To support null handling, Pinot must store null values in segments. The forward index stores the default value for null rows whether null storing is enabled or _not_.
```

## Default behavior

Pinot always stores column values in a [forward index](../../basics/indexing/forward-index.md).

> Forward index never stores null values but have to store a value for each row.

Reviewer: if Pinot always stores column values in a forward index, how is null value represented if forward index never stores null values?

Author: It is explained in the line above. Do you think it is not correctly explained?

Therefore independent of the null handling configuration, Pinot always stores a default value for nulls rows in the forward index.
The default value used in a column can be specified in the [schema](../../configuration-reference/schema.md) 
configuration by setting the `defaultNullValue` field spec.

{% hint style="info" %}
Remember that in the JSON used as table configuration, `defaultNullValue` must always be a String.

> If the column type is not String, Pinot will convert that value to the column type automatically.

Reviewer: ah, i see -- maybe in line 29-30 we could say something like Pinot requires a value for each row, because null values cannot be stored in a forward index, you must specify a `defaultNullValue` in table config?

Author: You don't must to specify `defaultNullValue`. There is a _default_ `defaultNullValue`, which depends on the actual type we are using. For example for integers it is the minimum value that can be referenced in C2. I think it would be interesting to include a table here that indicates, for each type, the `defaultNullValue` used by default. I'll try to do that for the next version

{% endhint %}

## Store nulls at ingestion time

In order to support any kind of null handling, the first thing that must be done is to store null values in the segments.
As said above, independently on whether null storing is enabled or not, the forward index will always store the default

> value for null rows.

Reviewer: whether null storing is enabled--what is the exact config? can we link to it here?

Author: I don't understand the suggestion. It is saying the complete opposite of what we the original text wanted to say.

> whether null storing is enabled--what is the exact config? can we link to it here?

That is what we are trying to explain in this specific section

Author: Well, not the opposite, but I think it is important to make clear that for null values, forward index always store the default value, independently of whether null storing is enabled or not.

Reviewer: yes, gotcha, thank you, @gortiz

> When null storing is enabled, Pinot will create a new index called the _null index_ or _null vector index_.

Reviewer: ```suggestion
When null storing is enabled, Pinot creates a new index called the `_null index_` or `_null vector index_`.
```

> This index stores the document ids of the rows that have null values for the column.

Reviewer: ```suggestion
This index stores the document IDs of the rows that have null values for the column.
```

{% hint style="danger" %}
Although storing null handling can be enabled after data has been ingested, data ingested before this mode is enabled

> will not store the null index and therefore it will be treated as not null.

Reviewer: ```suggestion
Although null handling can be enabled after data has been ingested, data ingested before null handling is enabled will not store the null index, so isn't treated as a null value.
```

Reviewer: not sure we need to say this line 49-- may be more explicit to say something like:
to ensure null values are properly stored, you must enable null handling before ingesting data.

Author: I think this may be a common scenario, so explicitly say what happens in this case is important.

Anyway I'm changing `storing null handling ` to `null storing`. Previous version used _null handling_ in both ingestion and query time and in this version I'm trying to use _null storing_ when talking about the ingestion part and _null handling_ when talking about querying to try to simplify the terms

Reviewer: Thank you, will add these terms to include in our glossary

{% endhint %}

Null storing support is defined per table. 

> This means that one table may be configured to store nulls while another table may not.

Reviewer: ```suggestion
Null support is configured per table. You can configure one table to store nulls, and configure another table to not store nulls.
```

There are two ways to define null storing support in Pinot:

1. [Column based null handling](#column-based-null-handling), where each column inside the table can be configured as 
nullable or not nullable. 

> This is the recommended way to enable null storing support and only way to support null handling in 

Reviewer: ```suggestion
1. [Column based null handling](#column-based-null-handling), where each column in a table is configured as nullable or not nullable. 
We recommend enabling null storing support by column. This is the only way to support null handling in the 
```

[multi-stage query engine](../../reference/multi-stage-engine.md).
2. [Table based null handling](#table-based-null-handling), where all columns in the table are considered nullable.

> This was the only way to handle null values in Pinot before 1.1.0, but it is considered deprecated.

Reviewer: ```suggestion
This is how null values were handled before Pinot 1.1.0 and now deprecated.
```

Reviewer: if we consider this deprecated, why not take it out here and say something about needing to update null configurations in the release notes?

Author: Because even it is deprecated, it is still supported. Which means they don't _have_ to migrate to the new way in order to store nulls.

Given the new alternative has been recently released, all customers are using right now this deprecated version. It may look strange for them to look for the way they are using right now and not see any information here. But we can move it somewhere else if you think it would be cleaner

### Column based null storing

Column based null storing was introduced in Pinot 1.1.0 and since then it is the recommended way to configure null 
storing in Pinot given it allows you to specify null storing at column basis.

> It is also the only way to support null handling in multi-stage query engine.

Reviewer: ```suggestion
We recommend configuring column based null storing, which lets you specify null handling per column and supports null handling in the multi-stage query engine.
```

In order to enable column based null handling, you have to set 
[enableColumnBasedNullHandling](../../configuration-reference/schema.md#Schema) to `true` in the schema configuration

> before ingesting the data. 

Reviewer: ```suggestion
To enable column based null handling, set [enableColumnBasedNullHandling](../../configuration-reference/schema.md#Schema) to `true` in the schema configuration before ingesting data. 
```

> Then you have to specify which columns are not nullable using the `notNull` field spec, which defaults to false.

Reviewer: ```suggestion
2. Specify which columns are not nullable using the `notNull` field spec, which defaults to false.
```

Reviewer: by not nullable, do we mean the column cannot handle nulls? what happens if the column cannot handle nulls and a null value is ingested?

Author: I think that can be inferred by the rest of the document:
- Whenever we receive a null value
  - Independent on whether we expect nulls or not, the default value for null (defined as `defaultNullValue` in the field inside the schema) is stored.
  - If nulls are expected for the column, we store that this cell is null in the null value index.

Therefore if we ingest a null value in a not nullable column, the default value is stored in the forward index and we treat it as a not null value

```json
{
  "schemaName": "my_table",
  "enableColumnBasedNullHandling": true,
  "dimensionFieldSpecs": [
    {
      "name": "notNullColumn",
      "dataType": "STRING",
      "notNull": true
    },
    {
      "name": "explicitNullableColumn",
      "dataType": "STRING",
      "notNull": false
    },
    {
      "name": "implicitNullableColumn",
      "dataType": "STRING"
    }
  ]
}
```

### Table based null storing

This is the only way to enable null storing in Pinot before 1.1.0, but it is deprecated since then.
Table based null storing is more expensive in terms of disk space and query performance than column based null storing.

> Also, it is not possible to support null handling in multi-stage query engine using table based null storing.

Reviewer: maybe we remove this section from latest, and include it in earlier versions of the docs (pre 1.1.0 release)?

Author: We definitively can use this new text to improve pre 1.1.0 release documentation. But as I said in another comment, even this mode is deprecated, it is still valid and for sure the most used way to configure nullability. Therefore I would not remove it.

Reviewer: i think i'm still figuring out the versioning process for Pinot (and know y'all have been working to figure that out!).

ideally, we'll get to a place we can have [Timeless documentation](https://developers.google.com/style/timeless-documentation), so our docs match whichever Pinot OSS version is selected in context 

![image](https://github.com/pinot-contrib/pinot-docs/assets/47797004/b9d689f0-489a-484c-8913-adc7e3d15713)

In order to enable table based null storing, you have to enable the `nullHandlingEnabled` configuration in 
[tableIndexConfig.nullHandlingEnabled](../../configuration-reference/table#table-index-config) before ingesting the data.

> Once that is done, all columns of the table are considered nullable.

Reviewer: ```suggestion
To enable table based null storing, enable the `nullHandlingEnabled` configuration in 
[tableIndexConfig.nullHandlingEnabled](../../configuration-reference/table#table-index-config) before ingesting data.
All columns in the table are now nullable.
```

{% hint style="warning" %}
Remember `nullHandlingEnabled` table configuration enables table based null handling

> while `enableNullHandling` is the query option that enables advanced null handling at query time.

Reviewer: can we add a use case example of when you'd use one or the other or both?

Author: It is explained you always need `nullHandlingEnabled` (or the new alternative to have column base null storing). When to enable `enableNullHandling` is explained above

Author: Adding a `See [advanced null handling support](#advanced-null-handling-support) for more information.`

Reviewer: thank you, @gortiz 

i think i got confused between 
`nullHandlingEnabled` and `enableNullHandling`  
and think which is `QueryTimeNullHandling` or `TableBasedNullHandling` or `ColumnBasedNullHandling`

}
```

## Null handling at query time

> In order to use null handling at query time, it is first necessary to enable [store nulls at ingestion time](#store-nulls-at-ingestion-time).
> Then basic null handling will be enabled by default and advanced null handling support can be optionally enabled.

Reviewer: ```suggestion
To enable basic null handling by at query time, enable Pinot to [store nulls at ingestion time](#store-nulls-at-ingestion-time).
Advanced null handling support can be optionally enabled.
```

{% hint style="warn" %}

> Remember that multi-stage query engine requires column based null storing.
> Tables that use table based null storing will be considered not nullable when using multi-stage query engine.

Reviewer: ```suggestion
The multi-stage query engine requires column based null storing. Tables with table based null storing are considered not nullable.
```

{% hint style="warn" %}
Even they have similar names, the `nullHandlingEnabled` table configuration and the `enableNullHandling` query option are different.
Remember `nullHandlingEnabled` table configuration modifies how segments are stored and `enableNullHandling` query option modifies how queries are executed.
{% endhint %}

When this `enableNullHandling` option is set to true, Pinot query engine uses a different execution path that interprets

> nulls in a standard SQL way.

Reviewer: ```suggestion
When the `enableNullHandling` option is set to `true`, the Pinot query engine uses a different execution path that interprets nulls in a standard SQL way.
```

This means that `IS NULL` and `IS NOT NULL` predicates will evaluate to `true` or `false` according to whether a null 
is detected (like in basic null support mode) but also aggregation functions like `COUNT`, `SUM`, `AVG`, `MODE`, etc. 
will deal with null values as expected (usually ignoring null values).

> In this mode, some indexes may not be usable, and queries may be significantly more expensive.

Reviewer: do we know which indexes may not be usable and/or other criteria that affects this?

Author: No, we don't have that information. The list may be too long and we would need to verify operator by operator. The list may also change on each version. I know it is not very informative but this is the best we have right now.

Performance degradation will impact all the columns in the table, even if columns that intervene in the query do not 

> actually contain null values.

Reviewer: ```suggestion
Performance degradation impacts all the columns in the table, including columns in the query do not contain null values.
```

Reviewer: not sure if this is suggested accurate (or what intervene meant in this context)

Author: Maybe
> Performance degradation impacts all the columns in the table, including columns in the query *that* do not contain null values.

?

Author: Changed to 
> Performance degradation impacts all the columns in the table, including columns in the query that do not contain null values.

> This happens even when table uses column base null storing.

Reviewer: Performance degradation happens?

Author: Yep

Author: Add reiteration to be more clear
